README.md

Truffler
========

Access web pages programmatically with [PhantomJS][phantom], for running tests or scraping information.

[![NPM version][shield-npm]][info-npm]
[![Node.js version support][shield-node]][info-node]
[![Build status][shield-build]][info-build]
[![Dependencies][shield-dependencies]][info-dependencies]
[![MIT licensed][shield-license]][info-license]

```js
var truffler = require('truffler');

var test = truffler(function(browser, page, done) {
    // the test function to run in PhantomJS
    done();
});

test.run('http://www.nature.com/', function(error, results) {
    console.log(results);
});
```


Table Of Contents
-----------------

- [Install](#install)
- [Usage](#usage)
- [Options](#options)
- [Examples](#examples)
- [Contributing](#contributing)
- [Migrating](#migrating)
- [License](#license)


Install
-------

Install Truffler with [npm][npm]:

```sh
npm install truffler
```


Usage
-----

Require in Truffler:

```js
var truffler = require('truffler');
```

Create a test runner by initialising Truffler with a test function. This test function has access to a PhantomJS [browser][phantom-browser] and [page][phantom-page] instance. The test function must accept a third argument which is a callback:

```js
var test = truffler(function(browser, page, done) {
    // ... perform testing here ...
    done(error, results);
});
```

Within this function, you have access to all of the behaviour in PhantomJS.

You can also instantiate Truffler with some [default options](#options) if you wish. This allows you to change the way PhantomJS and your page is loaded:

```js
var test = truffler({
    // ... options go here ...
}, function(browser, page, done) {
    // ... perform testing here ...
});
```

The `test.run` function can then be used to run your test function against a URL:

```js
test.run('http://www.nature.com/', function(error, results) {
    // ...
});
```

The `error` and `results` parameters contain errors and results from the PhantomJS run against your page. The results can be any object you like. Here's an example test function which returns the page title if it has one, or errors if not.

```js
var test = truffler(function(browser, page, done) {
    page.evaluate(
        function() {
            return document.title;
        },
        function(error, title) {
            if (!title) {
                return done(new Error('The page has no title!'));
            }
            done(null, title);
        }
    );
});

test.run('http://www.nature.com/', function(error, title) {
    // ... do something with the error and title ...
});
```


Options
-------

Truffler has lots of options you can use to change the way PhantomJS runs, or the way your page is loaded. Options can be set either on the Truffler instance when it's created or the individual test runs. This allows you to set some defaults which can be overridden per-test:

```js
// Set the default Foo header to "bar"
var test = truffler({
    page: {
        headers: {
            Foo: 'bar'
        }
    }
}, function() { /* ... */ });

// Run a test with the Foo header set to "bar"
test.run('http://www.nature.com/', function(error, title) { /* ... */});

// Run a test with the Foo header overridden
test.run('http://www.nature.com/', {
    page: {
        headers: {
            Foo: 'hello'
        }
    }
}, function(error, title) { /* ... */});
```

Below is a reference of all the options that are available:

### `log` (object)

An object which implments the methods `debug`, `error`, and `info` which will be used to report errors and test information.

```js
truffler({
    log: {
        debug: console.log.bind(console),
        error: console.error.bind(console),
        info: console.info.bind(console)
    }
});
```

Each of these defaults to an empty function.

### `page.headers` (object)

A key-value map of request headers to send when testing a web page.

```js
truffler({
    page: {
        headers: {
            Cookie: 'foo=bar'
        }
    }
});
```

Defaults to an empty object.

### `page.settings` (object)

A key-value map of settings to add to the PhantomJS page. For a full list of available settings, see the [PhantomJS page settings documentation][phantom-page-settings].

```js
truffler({
    page: {
        settings: {
            loadImages: false,
            userName: 'nature',
            password: 'say the magic word'
        }
    }
});
```

Defaults to:

```js
{
    resourceTimeout: 30000,
    userAgent: 'truffler/<version>'
}
```

### `page.viewport` (object)

The viewport width and height in pixels. The `viewport` object must have both `width` and `height` properties.

```js
truffler({
    page: {
        viewport: {
            width: 320,
            height: 480
        }
    }
});
```

Defaults to:

```js
{
    width: 1024,
    height: 768
}
```

### `phantom` (object)

A key-value map of settings to initialise PhantomJS with. This is passed directly into the `phantom` module – [documentation can be found here][phantom-node-options]. You can pass PhantomJS command-line parameters in the `phantom.parameters` option as key-value pairs.

```js
truffler({
    phantom: {
        path: 'PATH_TO_PHANTOMJS_EXE',
        port: 1234,
        parameters: {
            'ignore-ssl-errors': 'true'
        }
    }
});
```

Defaults to an empty object. If `phantom.port` is not specified, a random available port will be used.

### `timeout` (number)

The maximum time (in milliseconds) that Truffler should run for. This timeout can sometimes be exceeded, if a long-running task has started within PhantomJS itself. This is rare, but you shouldn't rely on exact timing.

If the timeout is exceeded, the test function will callback with an error and no results.

```js
truffler({
    timeout: 1000
});
```

Defaults to `30000` (30 seconds).


Examples
--------

### Basic Example

Run Truffler on a URL and output the page title:

```
node example/basic
```

### Multiple Example

Use [async][async] to run Truffler on multiple URLs in series, and output the page titles:

```
node example/multiple
```


Contributing
------------

To contribute to Truffler, clone this repo locally and commit your code on a separate branch.

Please write unit tests for your code, and check that everything works by running the following before opening a pull-request:

```sh
make ci
```


Migrating
---------

If you're using Truffler 1.0 and wish to migrate to 2.0, we've written a [Migration Guide](MIGRATION.md) to help with that.

It's recommended that you migrate to 2.0 as soon as possible, but [1.0 is still available on the 1.x branch][1.x]. We'll be providing support for 1.0 for the forseeable future, but there will be no new feature development.


License
-------

Truffler is licensed under the [MIT][info-license] license.  
Copyright &copy; 2015, Springer Nature



[1.x]: https://github.com/springernature/truffler/tree/1.x
[async]: https://github.com/caolan/async
[npm]: https://npmjs.org/
[phantom]: http://phantomjs.org/
[phantom-browser]: http://phantomjs.org/api/phantom/
[phantom-node-options]: https://github.com/sgentle/phantomjs-node#functionality-details
[phantom-page]: http://phantomjs.org/api/webpage/
[phantom-page-settings]: http://phantomjs.org/api/webpage/property/settings.html

[info-dependencies]: https://gemnasium.com/springernature/truffler
[info-license]: LICENSE
[info-node]: package.json
[info-npm]: https://www.npmjs.com/package/truffler
[info-build]: https://travis-ci.org/springernature/truffler
[shield-dependencies]: https://img.shields.io/gemnasium/springernature/truffler.svg
[shield-license]: https://img.shields.io/badge/license-MIT-blue.svg
[shield-node]: https://img.shields.io/badge/node.js%20support-0.12–5-brightgreen.svg
[shield-npm]: https://img.shields.io/npm/v/truffler.svg
[shield-build]: https://img.shields.io/travis/springernature/truffler/master.svg


HISTORY.md

# History

## 2.1.1 (2016-02-09)

  * Update repository references to springernature
  * Update the license

## 2.1.0 (2015-12-14)

  * Add the ability to specify the path to the PhantomJS binary

## 2.0.1 (2015-11-09)

  * Support Node.js 5.x
  * Update dependencies

## 2.0.0 (2015-10-08)

  * Initial 2.0 release
  * Overhaul the API
  * Drop Node.js 0.10 support
  * Switch the PhantomJS bridge
  * Specify a default `resourceTimeout` option
  * Add a `timeout` option
  * See the [migration guide](https://github.com/springernature/truffler/blob/master/MIGRATION.md#migrating-from-10-to-20) for details

## 1.0.5 (2015-10-01)

  * Add Node.js 4.x support
  * Update dependencies

## 1.0.4 (2015-09-08)

  * Update dependencies
  * Add code coverage reporting

## 1.0.3 (2015-07-06)

  * Update hasbin to fix an issue which broke Windows support

## 1.0.2 (2015-06-23)

  * Use JSCS to check for code-style issues
  * Update dependencies

## 1.0.1 (2015-05-20)

  * Add an example of testing multiple URLs

## 1.0.0 (2015-05-19)

  * Initial release


test/unit/setup.js
// jshint maxstatements: false
// jscs:disable disallowMultipleVarDecl, maximumLineLength
'use strict';

var assert = require('proclaim');
var mockery = require('mockery');
var sinon = require('sinon');

sinon.assert.expose(assert, {
	includeFail: false,
	prefix: ''
});

beforeEach(function() {
	mockery.enable({
		useCleanCache: true,
		warnOnUnregistered: false,
		warnOnReplace: false
	});
});

afterEach(function() {
	mockery.deregisterAll();
	mockery.disable();
});


test/unit/mock/hasbin.js
'use strict';

var sinon = require('sinon');

module.exports = {
	some: {
		sync: sinon.stub().returns(true)
	}
};


lib/truffler.js
'use strict';

var async = require('async');
var extend = require('node.extend');
var freeport = require('freeport');
var hasbin = require('hasbin');
var phantom = require('node-phantom-simple');
var pkg = require('../package.json');

module.exports = truffler;
module.exports.defaults = {
	log: {
		debug: /* istanbul ignore next */ function() {},
		error: /* istanbul ignore next */ function() {},
		info: /* istanbul ignore next */ function() {}
	},
	page: {
		headers: {},
		settings: {
			resourceTimeout: 30000,
			userAgent: 'truffler/' + pkg.version
		},
		viewport: {
			width: 1024,
			height: 768
		}
	},
	phantom: {},
	timeout: 30000
};

function truffler(options, testFunction) {
	return new Truffler(options, testFunction);
}

function Truffler(options, testFunction) {
	if (typeof options === 'function') {
		testFunction = options;
		options = {};
	}
	if (!((options && options.phantom && options.phantom.path) || isPhantomInstalled())) {
		throw new Error('PhantomJS binary was not found in PATH');
	}
	this.options = defaultOptions(module.exports.defaults, options);
	this.testFunction = testFunction;
}

Truffler.prototype.run = function(url, options, done) {
	done = arguments[arguments.length - 1];
	if (arguments.length === 2 && typeof url === 'string') {
		options = {};
	}
	if (arguments.length === 2 && typeof url !== 'string') {
		options = url;
		url = options.url;
	}
	this._run(url, options, done);
};

Truffler.prototype._run = function(url, options, done) {
	var state = {};
	state.testFunction = this.testFunction;
	url = this.sanitizeUrl(url);
	options = defaultOptions(this.options, options);

	// Start the timeout timer
	state.timer = setTimeout(function() {
		state.timeout = true;
	}, options.timeout);

	// Wrap functions to ensure timeout errors block further execution
	function guardAgainstTimeout(functionToGuard) {
		return function(next) {
			if (state.timeout) {
				return next(new Error('Truffler timed out (' + options.timeout + 'ms)'));
			}
			functionToGuard(next);
		};
	}

	// Start the testing
	async.series({

		// Randomize the port if one is not set
		randomizePort: guardAgainstTimeout(function(next) {
			if (options.phantom.port) {
				return next();
			}
			freeport(function(error, port) {
				if (error) {
					return next(error);
				}
				options.phantom.port = port;
				next();
			});
		}),

		// Create a PhantomJS browser
		createBrowser: guardAgainstTimeout(function(next) {
			phantom.create(options.phantom, function(error, browser) {
				if (error) {
					return next(error);
				}
				options.log.info('PhantomJS browser created');
				state.browser = browser;
				next();
			});
		}),

		// Create a PhantomJS page
		createPage: guardAgainstTimeout(function(next) {
			state.browser.createPage(function(error, page) {
				if (error) {
					return next(error);
				}
				options.log.debug('PhantomJS page created');
				state.page = page;
				next();
			});
		}),

		// Apply page headers
		applyPageHeaders: guardAgainstTimeout(function(next) {
			state.page.set('customHeaders', options.page.headers, next);
		}),

		// Apply page settings
		applyPageSettings: guardAgainstTimeout(function(next) {
			var settings = options.page.settings;
			var settingFunctions = Object.keys(settings).map(function(setting) {
				return state.page.set.bind(state.page, 'settings.' + setting, settings[setting]);
			});
			async.parallel(settingFunctions, next);
		}),

		// Apply page viewport size
		applyPageViewport: guardAgainstTimeout(function(next) {
			state.page.set('viewportSize', options.page.viewport, next);
		}),

		// Open the URL in PhantomJS
		openUrl: guardAgainstTimeout(function(next) {
			options.log.debug('Opening "' + url + '" in PhantomJS');
			state.page.open(url, function(error, status) {
				if (error) {
					return next(error);
				}
				if (status !== 'success') {
					return next(new Error('Page "' + url + '" could not be opened'));
				}
				options.log.debug('Opened "' + url + '" in PhantomJS');
				next();
			});
		}),

		// Run the test function
		runTestFunction: guardAgainstTimeout(function(next) {
			options.log.info('Testing the page "' + url + '"');
			state.testFunction(state.browser, state.page, function(error, result) {
				if (error) {
					options.log.error('Test function errored for "' + url + '"');
					return next(error);
				}
				options.log.debug('Test function ran for "' + url + '"');
				next(null, result);
			});
		})

	}, function(error, result) {
		clearTimeout(state.timer);
		if (state.page) {
			options.log.debug('PhantomJS page closed');
			state.page.close();
			state.page = null;
		}
		if (state.browser) {
			options.log.debug('PhantomJS browser exited');
			state.browser.exit();
			state.browser = null;
		}
		if (error) {
			return done(error);
		}
		done(null, result.runTestFunction);
	});
};

Truffler.prototype.sanitizeUrl = function(url) {
	url = url.trim();
	if (!/^[a-z]+:\/\//i.test(url)) {
		url = 'http://' + url;
	}
	return url;
};

function isPhantomInstalled() {
	return hasbin.some.sync(['phantomjs', 'phantomjs.exe']);
}

function defaultOptions(defaults, options) {
	return extend(true, {}, defaults, options);
}
